cohortextractor/generate_codelist_report.py

```diff
@@ -0,0 +1,109 @@
+import os
+
+import pandas as pd
+import structlog
+
+from .codelistlib import codelist_from_csv
+from .mssql_utils import mssql_dbapi_connection_from_url
+from .tpp_backend import make_batches_of_insert_statements, standardise_if_date
+
+logger = structlog.get_logger()
+
+
+def generate_codelist_report(output_dir, codelist_path, start_date, end_date):
+    """
+    Generate a pair of CSV files reporting usage of codes in given codelist.
+
+        * counts_per_code counts how many times each code appears in the given timeframe
+        * counts_per_week counts how many times all the codes appear in each week in the
+            given timeframe
+
+    Weeks begin on a Monday, and a code appearing on (say) a Wednesday is counted as
+    belonging to the following Monday.
+    """
+
+    codelist = codelist_from_csv(codelist_path, "snomedct")
+    start_date = standardise_if_date(start_date)
+    end_date = standardise_if_date(end_date)
+
+    queries = codelist_queries(codelist) + [
+        events_query(start_date, end_date),
+        population_query(start_date, end_date),
+        results_query(),
+    ]
+
+    connection = mssql_dbapi_connection_from_url(os.environ["DATABASE_URL"])
+    cursor = connection.cursor()
+
+    for query in queries:
+        logger.debug(query)
+        cursor.execute(query)
+
+    counts = pd.DataFrame(list(cursor), columns=["code", "date", "num"])
+
+    counts_per_code = counts.groupby("code")["num"].sum()
+    counts_per_week = counts.groupby(pd.Grouper(key="date", freq="W-MON"))["num"].sum()
+
+    os.makedirs(output_dir, exist_ok=True)
+
+    counts_per_code.to_csv(output_dir / "counts_per_code.csv")
+    counts_per_week.to_csv(output_dir / "counts_per_week.csv")
+
+
+def codelist_queries(codelist):
+    # These are SNOMED-specific values
+    collation = "Latin1_General_BIN"
+    max_code_len = 18
+
+    queries = [
+        f"""
+        CREATE TABLE #codelist (
+          code VARCHAR({max_code_len}) COLLATE {collation},
+          category VARCHAR(MAX)
+        )
+        """
+    ]
+    rows = [[code, ""] for code in codelist]
+    queries += make_batches_of_insert_statements(
+        "#codelist", ("code", "category"), rows
+    )
+    return queries
+
+
+def events_query(start_date, end_date):
+    return f"""
+    SELECT * INTO #events FROM (
+        SELECT Patient_ID, ConceptID, ConsultationDate
+        FROM CodedEvent_SNOMED
+        INNER JOIN #codelist
+            ON CodedEvent_SNOMED.ConceptID = #codelist.code
+        WHERE ConsultationDate BETWEEN '{start_date}' AND '{end_date}'
```

Comment:
I'd be tempted to remove the `standardise_if_date()` calls on L26-27 and use the [`quote()`](https://github.com/opensafely-core/cohort-extractor/blob/d25a266387003f8733f9256de5cfa37428ee4adc/cohortextractor/tpp_backend.py#L3617-L3628) function here instead e.g.
```
WHERE ConsultationDate BETWEEN {quote(start_date)} AND {quote(end_date)}
```

Same for the `population_query()` function below as well.

Building SQL via string interpolation is a bit yucky at the best of times, but I feel slightly better about it if the quoting is done right at the point where its used so you can't wonder whether it's been applied.

Reply:
Our strings have hyphens in them, which can apparently cause issues with SQL Server. The standardise_if_date function removes them.

Good point about the use of string interpolation though. pyodbc (which I assume we're using here), supports the use of parameterised queries, so I will give that a go

Reply:
Oh sorry I should have said that the `quote()` function applies that date standardisation as well. (There's a whole [sorry story](https://github.com/opensafely-core/databuilder/issues/92) behind this that fortunately we have a proper solution to now in databuilder.)

Parameterised queries would be great (although they'll still need that date reformatting thing sadly).

Reply:
Oh, yes, sorry. I should've looked at the quote() function rather than making assumptions from the name. 

We don't seem to be making use of parameterised queries in the cohort extractor. Was there a reason for that or was it just faster+easier to use quote? I know they aren't suitable for all cases, so don't know if a standard approach was taken?

Reply:
Originally we were generating SQL as text files which were fed into a separate Microsoft utility to actually execute them against the database, so parameterisation wasn't an option. This was before we managed to work around the issues involved with using pyODBC to fetch large amounts of data from MSSQL.

And yes, it's obviously incorrect for the `quote()` function to be doing the date transformation as well. But we only discovered the date inconsistency issue just before we were about to publish the first paper and so this was the easiest way of making sure we didn't miss any uses of dates anywhere.

There's a _lot_ of baggage from the early days in cohortextractor I'm afraid!

Reply:
Thank you. That makes a lot of sense. Code archaeology always brings up some interesting stuff.

I'm assuming the Microsoft utility was SQL Sever Agent? That brings back some memories!